File: packages/js/src/types/registry.ts
/**
 * Type registry for artifact types
 *
 * Manages both built-in and custom artifact types, providing
 * type detection, TTL lookup, and registration functionality.
 */

import micromatch from 'micromatch'
import {
  type ArtifactType,
  BUILT_IN_TYPES,
  DEFAULT_TYPE,
  isBuiltInType,
} from './built-in.js'

/**
 * Options for creating a TypeRegistry
 */
export interface TypeRegistryOptions {
  /** Whether to include built-in types (default: true) */
  includeBuiltIn?: boolean
  /** Custom types to register */
  customTypes?: Record<string, Partial<ArtifactType>>
}

/**
 * Type registry for managing artifact types
 */
export class TypeRegistry {
  private types: Map<string, ArtifactType> = new Map()
  private patternCache: Map<string, string> = new Map()

  constructor(options: TypeRegistryOptions = {}) {
    const { includeBuiltIn = true, customTypes = {} } = options

    // Load built-in types
    if (includeBuiltIn) {
      for (const [name, type] of Object.entries(BUILT_IN_TYPES)) {
        this.types.set(name, type)
      }
    }

    // Load custom types
    for (const [name, partialType] of Object.entries(customTypes)) {
      this.register({
        ...DEFAULT_TYPE,
        ...partialType,
        name,
      })
    }
  }

  /**
   * Get a type by name
   *
   * Custom types take precedence over built-in types with the same name.
   *
   * @param name - Type name
   * @returns The type or undefined
   */
  get(name: string): ArtifactType | undefined {
    return this.types.get(name)
  }

  /**
   * Register a custom type
   *
   * @param type - The type to register
   */
  register(type: ArtifactType): void {
    this.types.set(type.name, type)
    // Clear pattern cache when types change
    this.patternCache.clear()
  }

  /**
   * Unregister a type
   *
   * Note: Built-in types can be unregistered but this is not recommended.
   *
   * @param name - Type name to unregister
   * @returns true if the type was removed
   */
  unregister(name: string): boolean {
    const result = this.types.delete(name)
    if (result) {
      this.patternCache.clear()
    }
    return result
  }

  /**
   * Check if a type exists
   *
   * @param name - Type name
   * @returns true if the type exists
   */
  has(name: string): boolean {
    return this.types.has(name)
  }

  /**
   * List all registered types
   *
   * @returns Array of all types
   */
  list(): ArtifactType[] {
    return Array.from(this.types.values())
  }

  /**
   * List all type names
   *
   * @returns Array of type names
   */
  listNames(): string[] {
    return Array.from(this.types.keys())
  }

  /**
   * Detect the type of a file based on its path
   *
   * Types are checked in order of specificity (most specific patterns first).
   * Returns 'default' if no type matches.
   *
   * @param filePath - File path to check
   * @returns The detected type name
   */
  detectType(filePath: string): string {
    // Check cache first
    const cached = this.patternCache.get(filePath)
    if (cached) {
      return cached
    }

    // Normalize path separators
    const normalizedPath = filePath.replace(/\\/g, '/')

    // Check each type's patterns
    // Sort by pattern specificity (longer patterns first)
    const sortedTypes = this.list().sort((a, b) => {
      const aMaxLen = Math.max(...a.patterns.map((p) => p.length))
      const bMaxLen = Math.max(...b.patterns.map((p) => p.length))
      return bMaxLen - aMaxLen
    })

    for (const type of sortedTypes) {
      if (type.name === 'default') continue

      for (const pattern of type.patterns) {
        if (micromatch.isMatch(normalizedPath, pattern)) {
          this.patternCache.set(filePath, type.name)
          return type.name
        }
      }
    }

    // No match, return default
    this.patternCache.set(filePath, 'default')
    return 'default'
  }

  /**
   * Get the TTL for a file based on its type
   *
   * @param filePath - File path
   * @param override - Optional TTL override
   * @returns TTL in seconds
   */
  getTtl(filePath: string, override?: number): number {
    if (override !== undefined) {
      return override
    }

    const typeName = this.detectType(filePath)
    const type = this.get(typeName) || DEFAULT_TYPE
    return type.defaultTtl
  }

  /**
   * Get archive configuration for a file
   *
   * @param filePath - File path
   * @returns Archive configuration or null if archiving is disabled
   */
  getArchiveConfig(filePath: string): {
    afterDays: number
    storage: 'local' | 'cloud' | 'drive'
  } | null {
    const typeName = this.detectType(filePath)
    const type = this.get(typeName) || DEFAULT_TYPE

    if (type.archiveAfterDays === null || type.archiveStorage === null) {
      return null
    }

    return {
      afterDays: type.archiveAfterDays,
      storage: type.archiveStorage,
    }
  }

  /**
   * Get all files that match a type's patterns
   *
   * @param typeName - Type name
   * @param files - Array of file paths to filter
   * @returns Filtered file paths
   */
  filterByType(typeName: string, files: string[]): string[] {
    const type = this.get(typeName)
    if (!type) {
      return []
    }

    return files.filter((file) => {
      const normalized = file.replace(/\\/g, '/')
      return type.patterns.some((pattern) => micromatch.isMatch(normalized, pattern))
    })
  }

  /**
   * Check if a type is a built-in type
   *
   * @param name - Type name
   * @returns true if it's a built-in type
   */
  isBuiltIn(name: string): boolean {
    return isBuiltInType(name)
  }

  /**
   * Get the number of registered types
   */
  get size(): number {
    return this.types.size
  }

  /**
   * Clear pattern cache
   *
   * Call this if you need to force re-detection of types.
   */
  clearCache(): void {
    this.patternCache.clear()
  }
}

/**
 * Create a default TypeRegistry with built-in types
 *
 * @returns A new TypeRegistry instance
 */
export function createDefaultRegistry(): TypeRegistry {
  return new TypeRegistry()
}

File: packages/python/fractary_codex/types/registry.py
"""
Type registry for managing artifact types.

This module provides the TypeRegistry class for managing artifact types,
including built-in types and custom types from configuration.
"""

import fnmatch
from typing import Iterator, Optional

from .builtin import BUILT_IN_TYPES, DEFAULT_TTL, ArtifactType, get_all_built_in_types


class TypeRegistry:
    """Registry for managing artifact types.

    The registry maintains a collection of artifact types and provides
    methods for looking up types by name or file path pattern.

    Types are matched in priority order (highest first), so more specific
    types should have higher priority values.

    Examples:
        >>> registry = TypeRegistry()
        >>> ttl = registry.get_ttl("docs/api.md")
        >>> print(f"Cache for {ttl} seconds")

        >>> # Add custom type
        >>> registry.register(ArtifactType(
        ...     name="api-docs",
        ...     patterns=["docs/api/**/*.md"],
        ...     ttl=3600,
        ...     priority=100,  # Higher than built-in docs
        ... ))
    """

    def __init__(self, *, include_builtins: bool = True) -> None:
        """Initialize the type registry.

        Args:
            include_builtins: Whether to include built-in types (default: True)
        """
        self._types: dict[str, ArtifactType] = {}
        self._sorted_types: list[ArtifactType] | None = None

        if include_builtins:
            for type_def in BUILT_IN_TYPES.values():
                self._types[type_def.name] = type_def

    def register(self, type_def: ArtifactType) -> None:
        """Register a new artifact type.

        If a type with the same name already exists, it will be replaced.

        Args:
            type_def: Artifact type definition to register
        """
        self._types[type_def.name] = type_def
        self._sorted_types = None  # Invalidate cache

    def unregister(self, name: str) -> bool:
        """Unregister an artifact type by name.

        Args:
            name: Name of the type to unregister

        Returns:
            True if type was removed, False if not found
        """
        if name in self._types:
            del self._types[name]
            self._sorted_types = None
            return True
        return False

    def get(self, name: str) -> Optional[ArtifactType]:
        """Get an artifact type by name.

        Args:
            name: Type name

        Returns:
            ArtifactType if found, None otherwise
        """
        return self._types.get(name)

    def get_ttl(self, path: str) -> int:
        """Get the TTL for a file path based on matching types.

        This method checks all registered types in priority order and
        returns the TTL of the first matching type.

        Args:
            path: File path to match

        Returns:
            TTL in seconds (DEFAULT_TTL if no type matches)

        Examples:
            >>> registry = TypeRegistry()
            >>> registry.get_ttl("docs/api.md")
            86400
            >>> registry.get_ttl("config.yaml")
            3600
        """
        matched_type = self.match(path)
        return matched_type.ttl if matched_type else DEFAULT_TTL

    def match(self, path: str) -> Optional[ArtifactType]:
        """Find the artifact type that matches a file path.

        Types are checked in priority order (highest first).

        Args:
            path: File path to match

        Returns:
            Matching ArtifactType or None
        """
        # Normalize path for matching
        normalized_path = path.replace("\\", "/").lstrip("/")

        for type_def in self._get_sorted_types():
            if self._matches_patterns(normalized_path, type_def.patterns):
                return type_def

        return None

    def list_types(self) -> list[ArtifactType]:
        """List all registered types sorted by priority.

        Returns:
            List of ArtifactType sorted by priority descending
        """
        return list(self._get_sorted_types())

    def __iter__(self) -> Iterator[ArtifactType]:
        """Iterate over registered types in priority order."""
        return iter(self._get_sorted_types())

    def __len__(self) -> int:
        """Return the number of registered types."""
        return len(self._types)

    def __contains__(self, name: str) -> bool:
        """Check if a type name is registered."""
        return name in self._types

    def _get_sorted_types(self) -> list[ArtifactType]:
        """Get types sorted by priority (cached)."""
        if self._sorted_types is None:
            self._sorted_types = sorted(
                self._types.values(),
                key=lambda t: -t.priority,
            )
        return self._sorted_types

    @staticmethod
    def _matches_patterns(path: str, patterns: list[str] | tuple[str, ...]) -> bool:
        """Check if a path matches any of the patterns.

        Args:
            path: Normalized file path
            patterns: List of glob patterns

        Returns:
            True if any pattern matches
        """
        for pattern in patterns:
            # Handle ** for recursive matching
            if "**" in pattern:
                # fnmatch doesn't handle ** well, so we need special handling
                if _glob_match(path, pattern):
                    return True
            elif fnmatch.fnmatch(path, pattern):
                return True
        return False


def _glob_match(path: str, pattern: str) -> bool:
    """Match a path against a glob pattern with ** support.

    This provides basic ** support for patterns like:
    - docs/**/*.md - matches docs/foo.md and docs/bar/baz.md
    - **/*.yaml - matches foo.yaml and a/b/c.yaml
    """
    # Split pattern into parts
    pattern_parts = pattern.split("/")
    path_parts = path.split("/")

    return _match_parts(path_parts, pattern_parts)


def _match_parts(path_parts: list[str], pattern_parts: list[str]) -> bool:
    """Recursively match path parts against pattern parts."""
    if not pattern_parts:
        return not path_parts

    if not path_parts:
        # Remaining pattern must be all ** to match empty path
        return all(p == "**" for p in pattern_parts)

    pattern = pattern_parts[0]

    if pattern == "**":
        # ** can match zero or more path segments
        if len(pattern_parts) == 1:
            return True  # ** at end matches everything

        # Try matching ** against 0, 1, 2, ... path segments
        for i in range(len(path_parts) + 1):
            if _match_parts(path_parts[i:], pattern_parts[1:]):
                return True
        return False

    if fnmatch.fnmatch(path_parts[0], pattern):
        return _match_parts(path_parts[1:], pattern_parts[1:])

    return False


def create_default_registry() -> TypeRegistry:
    """Create a TypeRegistry with default built-in types.

    Returns:
        TypeRegistry with all built-in types registered
    """
    return TypeRegistry(include_builtins=True)
